<a href="https://colab.research.google.com/github/soumaya287/Machine-Learning/blob/main/KNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Handling Data

In [1]:
import csv

with open('iris.data.txt', 'r') as csvfile:

      lines = csv.reader(csvfile)

      for row in lines :

          print(', '.join(row))
      #dataset = list(lines)
      #print(len(dataset))
      

5.1, 3.5, 1.4, 0.2, Iris-setosa
4.9, 3.0, 1.4, 0.2, Iris-setosa
4.7, 3.2, 1.3, 0.2, Iris-setosa
4.6, 3.1, 1.5, 0.2, Iris-setosa
5.0, 3.6, 1.4, 0.2, Iris-setosa
5.4, 3.9, 1.7, 0.4, Iris-setosa
4.6, 3.4, 1.4, 0.3, Iris-setosa
5.0, 3.4, 1.5, 0.2, Iris-setosa
4.4, 2.9, 1.4, 0.2, Iris-setosa
4.9, 3.1, 1.5, 0.1, Iris-setosa
5.4, 3.7, 1.5, 0.2, Iris-setosa
4.8, 3.4, 1.6, 0.2, Iris-setosa
4.8, 3.0, 1.4, 0.1, Iris-setosa
4.3, 3.0, 1.1, 0.1, Iris-setosa
5.8, 4.0, 1.2, 0.2, Iris-setosa
5.7, 4.4, 1.5, 0.4, Iris-setosa
5.4, 3.9, 1.3, 0.4, Iris-setosa
5.1, 3.5, 1.4, 0.3, Iris-setosa
5.7, 3.8, 1.7, 0.3, Iris-setosa
5.1, 3.8, 1.5, 0.3, Iris-setosa
5.4, 3.4, 1.7, 0.2, Iris-setosa
5.1, 3.7, 1.5, 0.4, Iris-setosa
4.6, 3.6, 1.0, 0.2, Iris-setosa
5.1, 3.3, 1.7, 0.5, Iris-setosa
4.8, 3.4, 1.9, 0.2, Iris-setosa
5.0, 3.0, 1.6, 0.2, Iris-setosa
5.0, 3.4, 1.6, 0.4, Iris-setosa
5.2, 3.5, 1.5, 0.2, Iris-setosa
5.2, 3.4, 1.4, 0.2, Iris-setosa
4.7, 3.2, 1.6, 0.2, Iris-setosa
4.8, 3.1, 1.6, 0.2, Iris-setosa
5.4, 3.4

In [2]:
import csv

import random
import pandas as pd

def loadDataset(filename,split, trainingSet=[] , testSet=[]):
  
    
    with open(filename, 'r') as csvfile:

       lines = csv.reader(csvfile)
       dataset = list(lines)

       for x in range(len(dataset)-1):

           for y in range(4):

                dataset[x][y] = float(dataset[x][y])

           if random.random() < split:

               
             
                   trainingSet.append(dataset[x])
               
              
           else:

           
                   testSet.append(dataset[x])
      

In [3]:
#We can test this function out with our iris dataset, as follows:

trainingSet=[]

testSet=[]

loadDataset('iris.data.txt', 0.66, trainingSet, testSet)

print ('Train: ' + repr(len(trainingSet)))

print ('Test: ' + repr(len(testSet)) )

Train: 106
Test: 43


# 2. Similarity
## **Euclidean Distance**

In [4]:
import math
def euclideanDistance(instance1, instance2, length):
    distance = 0
    for x in range(length):
        distance += pow((instance1[x] - instance2[x]), 2)
    return math.sqrt(distance)

In [5]:
data1 = [2, 2, 2, 'a']

data2 = [4, 4, 4, 'b']

distance = euclideanDistance(data1, data2, 3)

print('Distance: ' + repr(distance))

Distance: 3.4641016151377544


# 3. Neighbors

In [6]:
import operator

def getNeighbors(trainingSet, testInstance, k):
   
    distances = []
    length = len(testInstance)-1
    for x in range(len(trainingSet)):
        dist = euclideanDistance(testInstance, trainingSet[x], length)
        distances.append((trainingSet[x], dist))
    distances.sort(key = operator.itemgetter(1))
    neighbors = []
    for x in range(k):
        neighbors.append(distances[x][0])
    return neighbors

In [7]:
trainSet = [[2, 2, 2, 'a'], [4, 4, 4, 'b']]

testInstance = [5, 5, 5]

k = 1

neighbors = getNeighbors(trainSet, testInstance, 1)

print(neighbors)

[[4, 4, 4, 'b']]


# 4. Response

In [8]:
def getResponse(neighbors):
    # Creating a list with all the possible neighbors
    classVotes = {}
    for x in range(len(neighbors)):
        response = neighbors[x][-1]
        if response in classVotes:
            classVotes[response] += 1
        else:
            classVotes[response] = 1
    sortedVotes = sorted(classVotes.items(), key=operator.itemgetter(1), reverse=True)
    return sortedVotes[0][0]

In [9]:
neighbors = [[1,1,1,'a'], [2,2,2,'a'], [3,3,3,'b']]
response = getResponse(neighbors)
print(response)

a


# 5. Accuracy

In [10]:
def getAccuracy(testSet, predictions):
    correct = 0
    for x in range(len(testSet)):
        if testSet[x][-1] == predictions[x]:
            correct += 1
    return (correct/float(len(testSet))) * 100.0

In [11]:
testSet = [[1,1,1,'a'], [2,2,2,'a'], [3,3,3,'b']]

predictions = ['a', 'a', 'a']

accuracy = getAccuracy(testSet, predictions)

print(accuracy)

66.66666666666666


In [14]:
#applying the full algorithme on our dataset iris
trainingSet=[]
testSet=[]
split = 0.66
loadDataset('iris.data.txt', split, trainingSet, testSet)
   
predictions=[]
k = 3
for x in range(len(testSet)):
        neighbors = getNeighbors(trainingSet, testSet[x], k)
        result = getResponse(neighbors)
        predictions.append(result)
        #print('> predicted=' + repr(result) + ', actual=' + repr(testSet[x][-1]))
accuracy = getAccuracy(testSet, predictions)
print('Accuracy: ', accuracy)

Accuracy:  94.64285714285714


After applying the **KNN** algorithm on our dataset using the **euclidean distance** as our metric , we get an **accuracy** equal to **94.64**.

# 7. Choosing another distance metric
## **Manhattan Distance**:

In [12]:
def manhattanDistance(instance1, instance2, length):
    distance = 0
    for x in range(length):
        distance += abs((instance1[x] - instance2[x]))
    return (distance)

In [13]:
def getNeighbors_manhattanDistance(trainingSet, testInstance, k):
   
    distances = []
    length = len(testInstance)-1
    for x in range(len(trainingSet)):
        dist = manhattanDistance(testInstance, trainingSet[x], length)
        distances.append((trainingSet[x], dist))
    distances.sort(key = operator.itemgetter(1))
    neighbors = []
    for x in range(k):
        neighbors.append(distances[x][0])
    return neighbors

In [15]:
trainingSet=[]
testSet=[]
split = 0.66
loadDataset('iris.data.txt', split, trainingSet, testSet)
   
predictions=[]
k = 3
for x in range(len(testSet)):
        neighbors = getNeighbors_manhattanDistance(trainingSet, testSet[x], k)
        result = getResponse(neighbors)
        predictions.append(result)
        #print('> predicted=' + repr(result) + ', actual=' + repr(testSet[x][-1]))
accuracy = getAccuracy(testSet, predictions)
print('Accuracy: ', accuracy)

Accuracy:  97.61904761904762


Changing the distance metric from euclidean distance to manhattan distance has **improved** the accuracy. It becomes **97.61**.

Thus **Manhattan Distance** is preferred over the Euclidean distance metric.

We can conclude that **the metric** we use to compute distances plays an **important role** in KNN model.